nmagent/internal/wireserver.go

```diff
@@ -0,0 +1,123 @@
+package internal
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+const WirePrefix string = "/machine/plugins/?comp=nmagent&type="
+
+var _ http.RoundTripper = &WireserverTransport{}
+
+// WireserverResponse represents a raw response from Wireserver
+type WireserverResponse map[string]json.RawMessage
+
+// StatusCode extracts the embedded HTTP status code from the response from Wireserver
+func (w WireserverResponse) StatusCode() (int, error) {
+	if status, ok := w["httpStatusCode"]; ok {
+		var statusStr string
+		err := json.Unmarshal(status, &statusStr)
+		if err != nil {
+			return 0, fmt.Errorf("unmarshaling httpStatusCode from Wireserver: %w", err)
+		}
+
+		if code, err := strconv.Atoi(statusStr); err != nil {
+			return code, fmt.Errorf("parsing http status code from wireserver: %w", err)
+		} else {
+			return code, nil
+		}
+	}
+	return 0, fmt.Errorf("no httpStatusCode property returned in Wireserver response")
+}
+
+// WireserverTransport is an http.RoundTripper that applies transformation
+// rules to inbound requests necessary to make them compatible with Wireserver
+type WireserverTransport struct {
+	Transport http.RoundTripper
+}
+
+// RoundTrip executes arbitrary HTTP requests against Wireserver while applying
+// the necessary transformation rules to make such requests acceptable to
+// Wireserver
+func (w *WireserverTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// the original path of the request must be prefixed with wireserver's path
+	origPath := req.URL.Path
+	path := WirePrefix
+	if req.URL.Path != "" {
+		path += req.URL.Path[1:]
+	}
+
+	// the query string from the request must have its constituent parts (?,=,&)
+	// transformed to slashes and appended to the query
+	if req.URL.RawQuery != "" {
+		query := req.URL.RawQuery
+		query = strings.ReplaceAll(query, "?", "/")
+		query = strings.ReplaceAll(query, "=", "/")
+		query = strings.ReplaceAll(query, "&", "/")
+		path += "/" + query
+	}
+
+	req.URL.Path = path
+	// ensure that nothing has changed from the caller's perspective by resetting
+	// the URL
+	defer func() {
+		req.URL.Path = origPath
+	}()
+
+	// wireserver cannot tolerate PUT requests, so it's necessary to transform those to POSTs
+	if req.Method == http.MethodPut {
+		req.Method = http.MethodPost
+		defer func() {
+			req.Method = http.MethodPut
+		}()
+	}
+
+	// all POST requests (and by extension, PUT) must have a non-nil body
+	if req.Method == http.MethodPost && req.Body == nil {
+		req.Body = io.NopCloser(strings.NewReader(""))
+	}
+
+	// execute the request to the downstream transport
+	resp, err := w.Transport.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+	// we want to close this because we're going to replace it
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return resp, nil
+	}
```

Comment:
It doesn't seem like this path consumes body to EOF? Sorry if I keep bringing it up, but I find it in general to be better practice to consume all bytes, especially since these should be fairly small payloads (no great advantage to streaming at the moment IMO). When these are not consumed, its possible that tcp connections won't be reused properly.